### PR TITLE
Implement Center Sampling for FCOS

### PR DESCRIPTION
**Pull request by ElectronicElephant**
It is proposed in the paper and the code is mainly from `https://github.com/yqyao/FCOS_PLUS/blob/0d20ba34ccc316650d8c30febb2eb40cb6eaae37/maskrcnn_benchmark/modeling/rpn/fcos/loss.py#L42`.

It should increase AP by ~ 0.8.

I'm not quite sure about the License issue.

### CODE REVIEW

**Comment by hellock** [comment]
Thanks for the PR! However the license is of FCOS_PLUS repo is not compatible with Apache 2.0, thus we will reimplement the method based on the paper. If you would like to work on it, please let us know and it would be much appreciated.

**Comment by ElectronicElephant** [comment]
> 
> 
> Thanks for the PR! However the license is of FCOS_PLUS repo is not compatible with Apache 2.0, thus we will reimplement the method based on the paper. If you would like to work on it, please let us know and it would be much appreciated.

Okay, I'll try to rewrite the code on my own. BTW, should I open a new PR or I can just continue on this?

**Comment by ElectronicElephant** [comment]
I have hard reset the branch to the master branch and I'll rewrite the code on my own. I will request a review on finish. It should not take much time, since the idea is quite simple.

**Comment by ElectronicElephant** [comment]
Okay, I have rewritten the code in my style. I don't have enough computational resources to train a full model, but I let it run for several iterations and it should work.

@hellock please review.

BTW, I found that the FCOS in mmdetection shares a lot of code with the official one. Just wondering is it licensed?

Thanks.

**Comment by hellock** [comment]
FCOS in mmdetection is implemented based on the paper, and with the collaboration of original authors.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 1]
```diff
@@ -377,8 +384,44 @@ def fcos_target_single(self, gt_bboxes, gt_labels, points, regress_ranges):
...
 
-        # condition1: inside a gt bbox
-        inside_gt_bbox_mask = bbox_targets.min(-1)[0] > 0
+        if self.center_sample_radius:
+            # condition1: inside a `center bbox`
+            radius = self.center_sample_radius
+            center_xs = (gt_bboxes[..., 0] + gt_bboxes[..., 2]) / 2
+            center_ys = (gt_bboxes[..., 1] + gt_bboxes[..., 3]) / 2
+            center_gts = gt_bboxes.new_zeros(gt_bboxes.shape)
+            stride = center_xs.new_zeros(center_xs.shape)
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/fcos_head.py]: center_gts = torch.zeros_like(gt_bboxes)

File: mmdet/models/anchor_heads/fcos_head.py [hunk 2]
```diff
@@ -49,7 +49,8 @@ def __init__(self,
                      use_sigmoid=True,
                      loss_weight=1.0),
                  conv_cfg=None,
-                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True)):
+                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True),
+                 center_sample_radius=0):  # 0 for disable `center sampling`
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/fcos_head.py]: This is somewhat misleading. It may be better to add a parameter `center_sampling` to indicate whether or not using center sampling.

**Comment by ElectronicElephant** [comment]
@yhcao6 Thanks for your comments. The config based on R50 is just added.
```
configs\fcos\fcos_r50_caffe_fpn_gn_1x_4gpu_center_sampling.py
```

In my previous work, `Center Sampling` will increase the training speed a bit and increase mAP by ~0.8 on COCO. I'm sorry that I just don't have enough computational resources to test the new implementation these days.

**Comment by yhcao6** [comment]
@ElectronicElephant , 
Center Sampling increases `FCOS_r50_gn` from 36.7 to 37.8, here is the [20200310_000712.log](https://github.com/open-mmlab/mmdetection/files/4310666/20200310_000712.log).

File: configs/fcos/fcos_r50_caffe_fpn_gn_1x_4gpu_center_sampling.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    hooks=[
+        dict(type='TextLoggerHook'),
+        # dict(type='TensorboardLoggerHook')
+    ])
+# yapf:enable
+# runtime settings
+total_epochs = 12
+dist_params = dict(backend='nccl')
+log_level = 'INFO'
+work_dir = './work_dirs/fcos_r50_caffe_fpn_gn_1x_4gpu'
```
Comment (yhcao6) [inline configs/fcos/fcos_r50_caffe_fpn_gn_1x_4gpu_center_sampling.py]: Change the name of the config to `fcos_center_r50_caffe_fpn_gn_1x_4gpu.py` and `work_dir` to `fcos_center_r50_caffe_fpn_gn_1x_4gpu`.

**Comment by ElectronicElephant** [comment]
@yhcao6 Thank you for commenting and testing!

@hellock With everything done, this PR is ready to merge~

Cheers.

**Comment by ElectronicElephant** [comment]
> 
> 
> @ElectronicElephant ,
> Center Sampling increases `FCOS_r50_gn` from 36.7 to 37.8, here is the [20200310_000712.log](https://github.com/open-mmlab/mmdetection/files/4310666/20200310_000712.log).

Hello, just a simple question. I'm wondering how to enable `online evaluation` just like the one in the log? Should I change `workflow` into `[('train', 1), ('val', 1)]` ? But the config in the log only has `[('train', 1)]`?

**Comment by yhcao6** [comment]
> > @ElectronicElephant ,
> > Center Sampling increases `FCOS_r50_gn` from 36.7 to 37.8, here is the [20200310_000712.log](https://github.com/open-mmlab/mmdetection/files/4310666/20200310_000712.log).
> 
> Hello, just a simple question. I'm wondering how to enable `online evaluation` just like the one in the log? Should I change `workflow` into `[('train', 1), ('val', 1)]` ? But the config in the log only has `[('train', 1)]`?

**Comment by yhcao6** [comment]
> > @ElectronicElephant ,
> > Center Sampling increases `FCOS_r50_gn` from 36.7 to 37.8, here is the [20200310_000712.log](https://github.com/open-mmlab/mmdetection/files/4310666/20200310_000712.log).
> 
> Hello, just a simple question. I'm wondering how to enable `online evaluation` just like the one in the log? Should I change `workflow` into `[('train', 1), ('val', 1)]` ? But the config in the log only has `[('train', 1)]`?

You can add `--validate` in training scripts.

**Comment by ElectronicElephant** [comment]
> 
> 
> > > @ElectronicElephant ,
> > > Center Sampling increases `FCOS_r50_gn` from 36.7 to 37.8, here is the [20200310_000712.log](https://github.com/open-mmlab/mmdetection/files/4310666/20200310_000712.log).
> > 
> > 
> > Hello, just a simple question. I'm wondering how to enable `online evaluation` just like the one in the log? Should I change `workflow` into `[('train', 1), ('val', 1)]` ? But the config in the log only has `[('train', 1)]`?
> 
> You can add `--validate` in training scripts.

Okay, my bad. I found it later in the doc so I deleted the comment. Anyway, thanks!

File: mmdet/models/anchor_heads/fcos_head.py [hunk 4]
```diff
@@ -377,8 +386,44 @@ def fcos_target_single(self, gt_bboxes, gt_labels, points, regress_ranges):
...
+                                             y_mins, gt_bboxes[..., 1])
+            center_gts[..., 2] = torch.where(x_maxs > gt_bboxes[..., 2],
+                                             gt_bboxes[..., 2], x_maxs)
+            center_gts[..., 3] = torch.where(y_maxs > gt_bboxes[..., 3],
+                                             gt_bboxes[..., 3], y_maxs)
+
+            left_cb = xs - center_gts[..., 0]
+            right_cb = center_gts[..., 2] - xs
+            top_cb = ys - center_gts[..., 1]
+            bottom_cb = center_gts[..., 3] - ys
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: May be renamed to `cb_dist_left`, `cb_dist_right`, etc.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 5]
```diff
@@ -49,7 +49,9 @@ def __init__(self,
                      use_sigmoid=True,
                      loss_weight=1.0),
                  conv_cfg=None,
-                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True)):
+                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True),
+                 center_sampling=False,
+                 center_sample_radius=1.5):
```
[the remark this answers is not in the document]
Comment (ElectronicElephant) [reply mmdet/models/anchor_heads/fcos_head.py]: Sure. Thanks.

File: mmdet/models/anchor_heads/fcos_head.py (hunk 4 above)
Comment (ElectronicElephant) [reply mmdet/models/anchor_heads/fcos_head.py]: Sure. Thanks.
